Path distance feature

silanus23 opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #5331  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | only unit tests|
| Does this PR contain AI generated software? |only numbers|
| Was this PR description generated by AI software? | No|

---

## Description of contribution in a few bullet points
Only path utils as descripted in the issue. It is finding nearest part in the path and current or waited position of robot. It interpolates between points cause the expected path is between points. There are 2 versions global version is reccomended and local version can sometimes fail. 
<!--

-->

## Description of documentation updates required from your changes

At this stage none needed.
<!--

-->

## Description of how this change was tested

I wrote diffirent trajectories. AI help came in here I used it to calculate the numbers. I have tested the local search with clover leaf and retracting windows too but it failed. This was actually expected as the local search is optimized version. I don't think those failings come from a logical issue but the limitations of the approach behind it. 
<!--

-->

---

## Future work that may be required in bullet points
Future side of the it is planned as you reccomend. Next one will be about creating a new msg type and publishing it from controller_server. 
<!--

-->

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

silanus23 commented on nav2_util/include/nav2_util/geometry_utils.hpp:
```diff
@@ -188,6 +188,57 @@ inline int find_next_matching_goal_in_waypoint_statuses(
+ * @param finish End point of target vector
+ * @return int
+ */
+inline double distanceToSegment(
```
Sorry didn't realize first time

silanus23 commented on nav2_util/include/nav2_util/geometry_utils.hpp:
```diff
@@ -188,6 +188,57 @@ inline int find_next_matching_goal_in_waypoint_statuses(
+ * @return int
+ */
+inline double distanceToSegment(
+  const geometry_msgs::msg::PoseStamped & point,
```
Changed inputs to those.

silanus23 commented on nav2_util/src/path_utils.cpp:
```diff
@@ -0,0 +1,129 @@
+  return result;
+}
+
+PathSearchResult distanceFromPath(
```
Yeap exactly like you said :) sorry couldn't catch this in first.

silanus23 commented on nav2_util/src/path_utils.cpp:
```diff
@@ -0,0 +1,129 @@
+  double min_distance = std::numeric_limits<double>::max();
+  size_t closest_segment = start_index;
+  size_t segments_checked = 0;
+  const size_t max_segments = static_cast<size_t>(search_window_length);
```
I am so sorry this PR costed you. I used helper tools and not copy-pasting I was thinking I was reviewing them enough. But seems I was mistaken. Again and again I am gratefull for your patience and valuable time. I can guess how hard it is for you to deliver this kind of detailed review. I hope this time I met this repo's standarts. I will be even more strict in my future contributions. Thanks.

mini-1235 commented on nav2_util/src/CMakeLists.txt:
In https://github.com/ros-navigation/navigation2/pull/5414, we drop the usage of nav_2d_msgs and nav_2d_utils outside dwb, if nav_2d_msgs is not used here, maybe we should remove it

silanus23 replied:
Thank yo so much for this warning. I will try to resolve it without them.

silanus23 commented:
Hello @stevemacnski First of all I am not trying to rush the proccess but to inform you about my current local progress so when you have time for my committed codes it will allign with your vision and we can move faster.  

About DWB Critics: I added the critics and a unit test. I have been able to create a case where that critic can create a diff here are the proof vids and parameters I used
![with_path_hug-ezgif com-video-to-gif-converter (1)](https://github.com/user-attachments/assets/f0522409-46f5-4782-be63-e69a9477701a)
![without_path_hug-ezgif com-video-to-gif-converter (1)](https://github.com/user-attachments/assets/bcf7a8e1-03aa-4eac-8418-cca00215ea66)
.

About Costmap Layer: I get results in here but incosistent. and to be honest I am not really sure about my expectations in here. First path is not getting a corridor but the second is answering. In addition sometimes I can not get response too. But when I do get response I think I get a decent corridor around the path ahead of robot. 

BT node: This started to work better with others giving less frequent fails and now able to recover faster.

MPPI critic: This one is a bit problematic as mppi is using diff kind of trajectory units. So I either have to convert it (which kills entire philosphy) or create an overloaded function that uses it's input types. I would like to hear your guidance if there is a better and efficient approach.

I missed 1 of your comments ( the 1 about doxygen of the util func) but I don't like too many tiny commits I will make it better in next commit.

I commit as you give feedback so I can be in synch with you and we can progress step by step. 

I tried to find places to utilize added functions. I think maybe distance_to_segment can be used in someplace but it can optimize some places if used in place euclidean distance or optimization in planner side of the repo . I am really not sure about it. Maybe create an rviz2 plugin

Thanks for your guidance and time from now.

SteveMacenski commented on nav2_controller/include/nav2_controller/controller_server.hpp:
```diff
@@ -236,6 +242,7 @@ class ControllerServer : public nav2::LifecycleNode
   std::unique_ptr<nav2_util::OdomSmoother> odom_sub_;
   std::unique_ptr<nav2_util::TwistPublisher> vel_publisher_;
   nav2::Subscription<nav2_msgs::msg::SpeedLimit>::SharedPtr speed_limit_sub_;
+  rclcpp::Publisher<nav2_msgs::msg::TrackingError>::SharedPtr tracking_error_pub_;
```
Change this to a `nav2::Publisher`

SteveMacenski commented on nav2_msgs/msg/TrackingError.msg:
```diff
@@ -0,0 +1,8 @@
+std_msgs/Header header
+float32 tracking_error
+uint32 last_index
+float32 cross_product
```
Is this useful?

SteveMacenski replied:
!

SteveMacenski commented on nav2_msgs/msg/TrackingError.msg:
```diff
@@ -0,0 +1,8 @@
+# Real-time tracking error for Nav2 controller
+
+std_msgs/Header header
+float32 tracking_error
```
Add a note about the positive vs negative for the left/right deviation from the path

SteveMacenski replied:
@silanus23 please do this :-)

SteveMacenski commented on nav2_msgs/msg/TrackingError.msg:
```diff
@@ -0,0 +1,8 @@
+
+std_msgs/Header header
+float32 tracking_error
+uint32 last_index
```
Maybe current_path_index?

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -500,6 +503,8 @@ void ControllerServer::computeControl()
 
       computeAndPublishVelocity();
 
+      publishTrackingState();
```
Lets also publish in the controller server feedback the tracking state message

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -750,6 +757,66 @@ void ControllerServer::updateGlobalPath()
+  nav2_msgs::msg::TrackingError tracking_error_msg;
+  tracking_error_msg.header.stamp = now();
+  tracking_error_msg.header.frame_id = robot_pose.header.frame_id;
+  tracking_error_msg.tracking_error = path_search_result.distance;
```
Shouldn't the sign of the distance be the side?

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -750,6 +757,66 @@ void ControllerServer::updateGlobalPath()
+  double cross_product = nav2_util::geometry_utils::cross_product_2d(
+    robot_pose_in_path_frame.pose.position, segment_start.pose, segment_end.pose);
+
+  nav2_msgs::msg::TrackingError tracking_error_msg;
```
Create as a unique pointer and then publish via `std::move()`. It is more optimal when using IPC or composition because the middleware doesn't need to copy the information into a pointer for communication use.

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -750,6 +757,66 @@ void ControllerServer::updateGlobalPath()
+  const double distance_to_goal = nav2_util::geometry_utils::euclidean_distance(
+    robot_pose, end_pose_in_robot_frame);
+
+  const auto path_search_result = nav2_util::distance_from_path(
```
Note that there's some related work in https://github.com/ros-navigation/navigation2/pull/5446 (FYI @mini-1235) which moves the path hanlders into the controller server. This will prune the global path automatically and as such provide a path that is already in the controller server frame (odom) and prune so that the first point on the path is the starting point of the robot. That would astly simplify the logic in this function and above

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -750,6 +757,66 @@ void ControllerServer::updateGlobalPath()
+    return;
+  }
+
+  const double distance_to_goal = nav2_util::geometry_utils::euclidean_distance(
```
We actualy publish the distance to goal already in the action feedback, I tihnk this can be removed + obtaining end pose in robot frame.

SteveMacenski reviewed (changes requested):
Very good!

SteveMacenski commented:
@silanus23 please don't force push - it actually makes it much harder for me to review since I can't see diffs of what changed. :-)

silanus23 commented:
@SteveMacenski I think I handled your requests. Added tracking error and publishing it there. I added velocity and distance to goal in tracking_error.msg too and deleted them from path follow to not duplicate. In advance the sign of the `tracking_error` is now deciding as you say.

sebatztian commented on nav2_msgs/msg/TrackingError.msg:
```diff
@@ -0,0 +1,8 @@
+# Real-time tracking error for Nav2 controller
```
should the msg name then not be something more general, i.e. TrackingFeedback or PoseTracking, if 4 of the 5 contents are not related to an actual error?

silanus23 commented:
I need sometime cause changing parts of the follow_path failing other things. I will convert it ready and ping you as soon as I handle the issues. Won't take much I guess.

silanus23 commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -667,41 +673,60 @@ void ControllerServer::computeAndPublishVelocity()
+      try {
+        // Compute path search result - always start from index 0
+        const auto path_search_result = nav2_util::distance_from_path(
+          current_path_, robot_pose_in_path_frame.pose, 0, search_window_);
```
Yeap that change probably slipped after trying to fix the force-push.😅

silanus23 commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -667,41 +673,60 @@ void ControllerServer::computeAndPublishVelocity()
-          closest_pose_idx = curr_idx;
+        // Always reset to 0 for next cycle (to start from beginning)
+        start_index_ = 0;
+
```
Seems plausible gonna try it probably try it and push it today.

sebatztian commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -692,21 +694,19 @@ void ControllerServer::computeAndPublishVelocity()
-        // Always reset to 0 for next cycle (to start from beginning)
-        start_index_ = 0;
+          current_path_, robot_pose_in_path_frame.pose, start_index_, search_window_);
 
```
update start_index_ to path_search_result.closest_segment_index, otherwise it stays at 0

mini-1235 commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -667,41 +675,59 @@ void ControllerServer::computeAndPublishVelocity()
+        tracking_error_msg->distance_to_goal = current_distance_to_goal_;
+        tracking_error_msg->speed = std::hypot(twist.linear.x, twist.linear.y);
+        tracking_error_msg->remaining_path_length =
+          nav2_util::geometry_utils::calculate_path_length(current_path_, start_index_);
```
Yes, the global path is still kept in `current_path_`, I added a new variable to keep the local path

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -267,6 +265,7 @@ ControllerServer::on_activate(const rclcpp_lifecycle::State & /*state*/)
 {
   RCLCPP_INFO(get_logger(), "Activating");
 
+  tracking_feedback_pub_->on_activate();
```
Please move with the other publisher activations, L277

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -693,37 +695,31 @@ void ControllerServer::computeAndPublishVelocity()
+      // Update current tracking error and publish
+      current_tracking_feedback_ = *tracking_feedback_msg;
+
+      tracking_feedback_pub_->publish(std::move(tracking_feedback_msg));
```
Check `pub->get_subscription_count() > 0` before publishing

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -234,8 +233,8 @@ ControllerServer::on_configure(const rclcpp_lifecycle::State & state)
   get_parameter("costmap_update_timeout", costmap_update_timeout_dbl);
-  tracking_error_pub_ = create_publisher<nav2_msgs::msg::TrackingErrorFeedback>("tracking_error",
-    10);
+  tracking_feedback_pub_ = create_publisher<nav2_msgs::msg::TrackingFeedback>("tracking_feedback"
+  );
```
```suggestion
  tracking_feedback_pub_ = create_publisher<nav2_msgs::msg::TrackingFeedback>("tracking_feedback");
```

SteveMacenski commented on nav2_util/include/nav2_util/path_utils.hpp:
```diff
@@ -38,7 +38,14 @@ struct PathSearchResult
 /**
  * @brief Finds the minimum distance from a robot pose to a path segment.
- *
+ *Especially considering we need to explain to users that if start index
```
Please actually read my comment and extract the information to write a comment. Obviously we shouldn't write "Especially considering we need to explain to users...." :wink:

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -66,6 +68,8 @@ ControllerServer::ControllerServer(const rclcpp::NodeOptions & options)
   declare_parameter("odom_topic", rclcpp::ParameterValue("odom"));
   declare_parameter("odom_duration", rclcpp::ParameterValue(0.3));
 
```
```suggestion
```

SteveMacenski commented on nav2_msgs/action/FollowPath.action:
```diff
@@ -22,6 +22,7 @@ std_msgs/Empty result
+# feedback definition
+# Real-time tracking error indicating which side of the path the robot is on
+# If the tracking_error is positive, the robot is to the left of the path
+nav2_msgs/TrackingFeedback tracking_feedback
```
Where is this feedback used currently in the Nav2 code? This changes so that needs to be adjusted too. I'm thinking the nav2_simple_commander API for `followPath` and the demos that use it perhaps

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -667,41 +675,55 @@ void ControllerServer::computeAndPublishVelocity()
   action_server_->publish_feedback(feedback);
 }
 
+
```
```suggestion
```

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -667,41 +675,55 @@ void ControllerServer::computeAndPublishVelocity()
-  nav_msgs::msg::Path & current_path = current_path_;
-  auto find_closest_pose_idx = [&robot_pose_in_path_frame, &current_path]()
+  double current_distance_to_goal_ = 0.0;
+  nav2_msgs::msg::TrackingFeedback current_tracking_feedback_;
```
```suggestion
  nav2_msgs::msg::TrackingFeedback current_tracking_feedback;
```

This is a local member, there shouldn't be an underscore

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -667,41 +675,55 @@ void ControllerServer::computeAndPublishVelocity()
+
+      // Update current tracking error and publish
+      current_tracking_feedback_ = *tracking_feedback_msg;
+
```
```suggestion
```

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -667,41 +675,55 @@ void ControllerServer::computeAndPublishVelocity()
-
-  nav_msgs::msg::Path & current_path = current_path_;
-  auto find_closest_pose_idx = [&robot_pose_in_path_frame, &current_path]()
+  double current_distance_to_goal_ = 0.0;
```
```suggestion
  double current_distance_to_goal = 0.0;
```

Local variable; no need for underscore

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -667,41 +675,55 @@ void ControllerServer::computeAndPublishVelocity()
+  double current_distance_to_goal_ = 0.0;
+  nav2_msgs::msg::TrackingFeedback current_tracking_feedback_;
+  // Compute and publish tracking error
+  if (!current_path_.poses.empty() && current_path_.poses.size() >= 2) {
```
```suggestion
  if (current_path_.poses.size() >= 2) {
```

the empty check is redundant

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -667,41 +675,55 @@ void ControllerServer::computeAndPublishVelocity()
+  double current_distance_to_goal_ = 0.0;
+  nav2_msgs::msg::TrackingFeedback current_tracking_feedback_;
+  // Compute and publish tracking error
+  if (!current_path_.poses.empty() && current_path_.poses.size() >= 2) {
+    // Transform robot pose to path frame
+    geometry_msgs::msg::PoseStamped robot_pose_in_path_frame;
+    if (nav2_util::transformPoseInTargetFrame(
+            pose, robot_pose_in_path_frame, *costmap_ros_->getTfBuffer(),
+            current_path_.header.frame_id, costmap_ros_->getTransformTolerance()))
     {
-      size_t closest_pose_idx = 0;
-      double curr_min_dist = std::numeric_limits<double>::max();
-      for (size_t curr_idx = 0; curr_idx < current_path.poses.size(); ++curr_idx) {
-        double curr_dist =
-          nav2_util::geometry_utils::euclidean_distance(robot_pose_in_path_frame,
-          current_path.poses[curr_idx]);
-        if (curr_dist < curr_min_dist) {
-          curr_min_dist = curr_dist;
-          closest_pose_idx = curr_idx;
-        }
+      // Transform goal pose to robot frame for distance calculation
+      geometry_msgs::msg::PoseStamped transformed_end_pose;
+      if (nav2_util::transformPoseInTargetFrame(
+            end_pose_, transformed_end_pose, *costmap_ros_->getTfBuffer(),
+            pose.header.frame_id, costmap_ros_->getTransformTolerance()))
+      {
+        current_distance_to_goal_ = nav2_util::geometry_utils::euclidean_distance(
+          pose, transformed_end_pose);
       }
-      return closest_pose_idx;
-    };
 
-  const std::size_t closest_pose_idx = find_closest_pose_idx();
-  feedback->distance_to_goal = nav2_util::geometry_utils::calculate_path_length(current_path_,
-      closest_pose_idx);
```
I'm looking at this block now and there needs to be 2 changes

* Before we threw an exception if the transformation failed that this skips https://github.com/ros-navigation/navigation2/blob/main/nav2_controller/src/controller_server.cpp#L676 Please redo this `if (!transformPoseInTargetFrame(...))` logic to remove the nesting to flatten out this code. 
* shouldn't this already be in the path frame as an element of the path itself? https://github.com/ros-navigation/navigation2/blob/bc84c558415cc01701c616f0c400f279108dd450/nav2_controller/src/controller_server.cpp#L688-L696 I don't think this both is required for `current_distance_to_goal`

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -268,8 +273,8 @@ ControllerServer::on_activate(const rclcpp_lifecycle::State & /*state*/)
   }
   vel_publisher_->on_activate();
   action_server_->activate();
```
Lets try this one last time, this needs to go between the vel publisher and the action server, so that its available by the time the action server is active, but in the same code block as all activations

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -228,6 +232,7 @@ ControllerServer::on_configure(const rclcpp_lifecycle::State & state)
 
   double costmap_update_timeout_dbl;
   get_parameter("costmap_update_timeout", costmap_update_timeout_dbl);
+  tracking_feedback_pub_ = create_publisher<nav2_msgs::msg::TrackingFeedback>("tracking_feedback");
```
Ditto here, put this below `vel_publisher_`. This is in the middle of the costmap update timeout param setting, I'm not sure why you'd put this here...

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -666,39 +672,49 @@ void ControllerServer::computeAndPublishVelocity()
   RCLCPP_DEBUG(get_logger(), "Publishing velocity at time %.2f", now().seconds());
   publishVelocity(cmd_vel_2d);
+  double current_distance_to_goal = 0.0;
+  nav2_msgs::msg::TrackingFeedback current_tracking_feedback;
```
```suggestion

  nav2_msgs::msg::TrackingFeedback current_tracking_feedback;
```

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -666,39 +672,49 @@ void ControllerServer::computeAndPublishVelocity()
-    throw nav2_core::ControllerTFError("Failed to transform robot pose to path frame");
+    // Transform goal pose to robot frame for distance calculation
+    geometry_msgs::msg::PoseStamped transformed_end_pose;
+
```
```suggestion
```

SteveMacenski commented on nav2_util/src/path_utils.cpp:
```diff
@@ -0,0 +1,85 @@
+
+  const auto & segment_start = path.poses[result.closest_segment_index];
+  const auto & segment_end = path.poses[result.closest_segment_index + 1];
+
+  const double cross_product = geometry_utils::cross_product_2d(
+    robot_pose.position, segment_start.pose, segment_end.pose);
+
+  result.distance *= (cross_product >= 0.0 ? 1.0 : -1.0);
```
```suggestion

  // Obtain the signed direction of the cross track error
  const double cross_product = geometry_utils::cross_product_2d(
    robot_pose.position, segment_start.pose, segment_end.pose);
  result.distance *= (cross_product >= 0.0 ? 1.0 : -1.0);
```

SteveMacenski reviewed (changes requested):
Should be good to go now. @mini-1235 @sebatztian any last review comments?

SteveMacenski commented:
@mini-1235 please let me know if all of your comments have been resolved

SteveMacenski commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -134,6 +137,7 @@ ControllerServer::on_configure(const rclcpp_lifecycle::State & state)
   get_parameter("failure_tolerance", failure_tolerance_);
   get_parameter("use_realtime_priority", use_realtime_priority_);
   get_parameter("publish_zero_velocity", publish_zero_velocity_);
+  get_parameter("search_window", search_window_);
```
OK - I just retriggered CI to see if it was a fluke. Do you see these tests failing locally as well?

mini-1235 commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -666,39 +673,49 @@ void ControllerServer::computeAndPublishVelocity()
-    throw nav2_core::ControllerTFError("Failed to transform robot pose to path frame");
+    // Transform goal pose to robot frame for distance calculation
+    geometry_msgs::msg::PoseStamped transformed_end_pose;
+    current_distance_to_goal = nav2_util::geometry_utils::euclidean_distance(
```
In `isGoalReached` we also transform to odom frame too, I think we can make it a member so that we only transform once

silanus23 commented on nav2_controller/src/controller_server.cpp:
```diff
@@ -666,39 +673,54 @@ void ControllerServer::computeAndPublishVelocity()
-    throw nav2_core::ControllerTFError("Failed to transform robot pose to path frame");
+    // Transform goal pose to robot frame for distance calculation
+    geometry_msgs::msg::PoseStamped transformed_end_pose;
+    if (!nav2_util::transformPoseInTargetFrame(
```
I am planning to handle them today. Thanks for these.
